src/modules/assessments/components/AssessmentForm.tsx

```diff
@@ -119,9 +117,30 @@ const AssessmentForm: React.FC<Props> = ({
     if (assessment && !assessment.inProgress) setLocked(true);
   }, [assessment, canEdit]);
 
+  // Choose the FormDefiniton to use for rendering, and filter it down based on client attributes (Data Collected About rules).
+  const definition = useMemo(() => {
+    // Logic for choosing a definition:
+    //   - When viewing an assessment ("locked"), we use the Form Definition that was most recently used to
+    //     submit the assessment. That gets passed in as the definition prop.
+    //   - When editing a WIP assessment, we also use the same Form Definition that was used to create it.
+    //   - When editing a non-WIP assessment, we use the most recent version of the Form Definition. If this differs
+    //     from the other definition, then 'upgradedDefinitionForEditing' is present on the Assessment.
+    let fd = definitionProp;
+    if (
+      !locked &&
+      !assessment?.inProgress &&
+      assessment?.upgradedDefinitionForEditing
+    ) {
+      fd = assessment.upgradedDefinitionForEditing;
+    }
+    // Apply "data collected about" rules to filter down the definition to relevant items
+    const relationshipToHoH = enrollment.relationshipToHoH;
+    return applyDefinitionRulesForClient(fd, client, relationshipToHoH);
```

Comment:
Calling `applyDefinitionRulesForClient` was previously peppered throughout the codebase, at the locations where the definition is fetched. Now, it is only called in one place, right here, closer to where the form rendering occurs.

I think this is easier to reason about. We don't have system test coverage for this, but I opened https://github.com/open-path/Green-River/issues/6506 -- I started working on a system test for this locally, but decided to de-scope it. QA will be needed on this front.